clients/src/main/java/org/apache/kafka/common/requests/RequestUtils.java

```diff
@@ -38,41 +37,23 @@ private RequestUtils() {}
             Optional.empty() : Optional.of(leaderEpoch);
     }
 
-    // visible for testing
-    public static boolean hasIdempotentRecords(ProduceRequest request) {
-        return flags(request).getKey();
-    }
-
-    // visible for testing
     public static boolean hasTransactionalRecords(ProduceRequest request) {
-        return flags(request).getValue();
+        return flag(request, RecordBatch::isTransactional);
     }
 
     /**
-     * Get both hasIdempotentRecords flag and hasTransactionalRecords flag from produce request.
-     * Noted that we find all flags at once to avoid duplicate loop and record batch construction.
-     * @return first flag is "hasIdempotentRecords" and another is "hasTransactionalRecords"
+     * find a flag from all records of a produce request.
+     * @param request produce request
+     * @param predicate used to predicate the record
+     * @return true if there is any matched flag in the produce request. Otherwise, false
      */
-    public static AbstractMap.SimpleEntry<Boolean, Boolean> flags(ProduceRequest request) {
-        boolean hasIdempotentRecords = false;
-        boolean hasTransactionalRecords = false;
-        for (ProduceRequestData.TopicProduceData tpd : request.data().topicData()) {
-            for (ProduceRequestData.PartitionProduceData ppd : tpd.partitionData()) {
-                BaseRecords records = ppd.records();
-                if (records instanceof Records) {
-                    Iterator<? extends RecordBatch> iterator = ((Records) records).batches().iterator();
-                    if (iterator.hasNext()) {
-                        RecordBatch batch = iterator.next();
-                        hasIdempotentRecords = hasIdempotentRecords || batch.hasProducerId();
-                        hasTransactionalRecords = hasTransactionalRecords || batch.isTransactional();
-                    }
-                }
-                // return early
-                if (hasIdempotentRecords && hasTransactionalRecords)
-                    return new AbstractMap.SimpleEntry<>(true, true);
-            }
-        }
-        return new AbstractMap.SimpleEntry<>(hasIdempotentRecords, hasTransactionalRecords);
+    static boolean flag(ProduceRequest request, Predicate<RecordBatch> predicate) {
+        return request.data().topicData().stream().flatMap(tpd -> tpd.partitionData().stream())
+                .map(ProduceRequestData.PartitionProduceData::records)
+                .filter(records -> records instanceof Records)
+                .map(r -> ((Records) r).batches().iterator())
+                .filter(Iterator::hasNext)
+                .anyMatch(iter -> predicate.test(iter.next()));
```

Comment:
Have we checked if this is not much slower than the for loop? Since this is in the produce path, we should be careful about unnecessary overhead.

Reply:
Especially as we have multiple stages in the stream pipeline.

Reply:
> Have we checked if this is not much slower than the for loop?

update a commit with benchmark (will be removed by final commit). You are right. We should take care performance in the produce path. will rewrite the code by pure for loop and the benchmark is shown below (according to https://github.com/apache/kafka/pull/9884/commits/dd08ac39fde6ea297f5a51c72d20a6d786f221e2)

```
ProducerRequestBenchmark.findFlagByForLoop                               avgt   15   103.732 ±   6.511   ns/op
ProducerRequestBenchmark.findFlagByForLoop:·gc.alloc.rate                avgt   15  4637.362 ± 284.903  MB/sec
ProducerRequestBenchmark.findFlagByForLoop:·gc.alloc.rate.norm           avgt   15   528.000 ±   0.001    B/op
ProducerRequestBenchmark.findFlagByForLoop:·gc.churn.G1_Eden_Space       avgt   15  4638.660 ± 279.858  MB/sec
ProducerRequestBenchmark.findFlagByForLoop:·gc.churn.G1_Eden_Space.norm  avgt   15   528.194 ±   5.364    B/op
ProducerRequestBenchmark.findFlagByForLoop:·gc.churn.G1_Old_Gen          avgt   15     0.004 ±   0.001  MB/sec
ProducerRequestBenchmark.findFlagByForLoop:·gc.churn.G1_Old_Gen.norm     avgt   15    ≈ 10⁻³              B/op
ProducerRequestBenchmark.findFlagByForLoop:·gc.count                     avgt   15   700.000            counts
ProducerRequestBenchmark.findFlagByForLoop:·gc.time                      avgt   15   469.000                ms
ProducerRequestBenchmark.findFlagByStream                                avgt   15   345.205 ±  12.776   ns/op
ProducerRequestBenchmark.findFlagByStream:·gc.alloc.rate                 avgt   15  3286.684 ± 124.034  MB/sec
ProducerRequestBenchmark.findFlagByStream:·gc.alloc.rate.norm            avgt   15  1248.000 ±   0.001    B/op
ProducerRequestBenchmark.findFlagByStream:·gc.churn.G1_Eden_Space        avgt   15  3289.274 ± 125.748  MB/sec
ProducerRequestBenchmark.findFlagByStream:·gc.churn.G1_Eden_Space.norm   avgt   15  1249.056 ±  17.331    B/op
ProducerRequestBenchmark.findFlagByStream:·gc.churn.G1_Old_Gen           avgt   15     0.008 ±   0.002  MB/sec
ProducerRequestBenchmark.findFlagByStream:·gc.churn.G1_Old_Gen.norm      avgt   15     0.003 ±   0.001    B/op
ProducerRequestBenchmark.findFlagByStream:·gc.count                      avgt   15   597.000            counts
ProducerRequestBenchmark.findFlagByStream:·gc.time                       avgt   15   377.000                ms
```